### ISSUE TITLE
`prefer-flat-map` being erroneously triggered

### ISSUE THREAD

**Issue by gunar** [jfmengels/eslint-plugin-lodash-fp#30]
The rule `prefer-flat-map` is being trigger with the following line and I don't know why.

```js
    const filterOut = _.flow(_.get('foods'), _.reject('tomato'), _.reject('strawberry'), _.reject({ id: 'SUGARID' }))
```

Thanks :balloon:

**Comment by jfmengels** [comment]
Hey @gunar!

Thanks for reporting, but I can't seem to reproduce this, could you give me an (minimal?) example with Lodash's import/require statement and all?

**Comment by gunar** [comment]
### file.js

```js
'use strict'

const _ = require('lodash/fp')
const filterOut = _.flow(_.get('foods'), _.reject('tomato'), _.reject('strawberry'), _.reject({ id: 'SUGARID' }))
filterOut()
```

### .eslintrc
```
{
  "extends": [
    "airbnb/base",
    "plugin:lodash-fp/recommended"
  ],
  "plugins": [ "lodash-fp" ],
  "parserOptions": {
    "sourceType": "script",
  },
```

### dependencies
```
    "eslint": "^2.10.2",
    "eslint-config-airbnb": "^6.2.0",
    "eslint-plugin-lodash-fp": "^1.2.0",
    "lodash": "^4.11.2",
```

**Comment by jfmengels** [comment]
Ha! Found it!

It's not actually the `prefer-flat-map` rule that has an error, but rather the `prefer-composition-grouping` rule that is triggered, and that it currently gives the same error message as `prefer-flat-map`, and the same problem exists for `no-extraneous-iteratee-args`. Ugh... shameful... That's what you guess when you develop the logic first and then the error message and have no review :smile:.

I'll fix that tonight so that both of the rules return a proper error message.

Then, to come back to your error, you have multiple of the same operations (among `map`, `filter` and `reject`) next to each other, which can be improved upon.

```js
_.flow(
  _.get('foods'),
  _.reject('tomato'),
  _.reject('strawberry'),
  _.reject({ id: 'SUGARID' })
)

// -->

_.flow(
  _.get('foods'),
  _.reject(v => v.tomato || v.strawberry || v.id === 'SUGARID')
)
```

I know that when you use some compiled functional languages, the former is compiled into the latter, as it is more efficient for processing, and it may in some cases be easier to read, especially if you give the function a nice name. That's why I proposed this rule. I did not know/think of the property shorthand though.

```js
_.flow(
  _.map('a'),
  _.map('b'),
  _.map('c'),
  _.filter('foo')
);
// vs
_.flow(
  _.map('a.b.c'),
  _.filter('foo')
);
// Much simpler and faster in this case for instance
```

See https://github.com/jfmengels/eslint-plugin-lodash-fp/blob/master/docs/rules/prefer-composition-grouping.md for more details.

**Comment by gunar** [comment]
Awesome, man. I'll test it as soon as you push it.

I'd ask and advise you to steer this project into code readability over performance. I really don't care about performance. Most JS projects don't.

**Comment by jfmengels** [comment]
> I'd ask and advise you to steer this project into code readability over performance. I really don't care about performance. Most JS projects don't.

Agreed. In this case, I think it's a *possible* win in both readability and performance. Then again, if people don't like it, they can turn it off.

And for future rules, my main objectives are:
- avoiding lodash fp related errors
- Port the useful rules from `eslint-plugin-lodash` (you've probably seen all the open issues for those)
- Guide users into writing more readable code, or more functional in a way

Performance is absolutely not my main goal, and I wouldn't even know what kind of other performance tips I could give, so no worries there ;)

**Comment by jfmengels** [comment]
Just published `1.3.0`. Let me know if it all works for you, thanks :)

**Comment by gunar** [comment]
Thank you. It doesn't quite solve for me since flow doesn't accept shorthand operators.

```js
reject(flow('tomato', 'strawberry'))
```

**Comment by jfmengels** [comment]
Actually, you can, using `overEvery` or `overSome`, as they accept shorthands. Just learned that today. Let me know if this helps you :)

```js
_.flow(
  _.get('foods'),
  _.reject(_.overSome([
    'tomato',
    'strawberry',
    {id: 'SUGARID'}
  ]))
)(data)
```

**Comment by gunar** [comment]
That's perfect. Thank you!

**Comment by gunar** [comment]
Can we make the error message suggest overSome or overEvery?

**Comment by jfmengels** [comment]
I was going to add it in the documentation of the rule, but we could suggest overSome/overEvery when using filter/reject (not for map, doesn't make much sense).

What do you think of?
- "Prefer regrouping successive calls of `' + operation.name + '` into one function or function call" for map (that is the current message, no change)
- "Prefer regrouping successive calls of `' + operation.name + '` into one function or function call. You might want to use `overEvery`" for filter, and same thing for reject but with `overSome` instead of `overEvery`?

**Comment by gunar** [comment]
My opinion is:

- "Prefer grouping successive calls of " + operation.name + " into one function or function call" for map 
- "Prefer grouping successive calls of " + operation.name + " into one function or function call (e.g. with overEvery/overSome)" for filter/reject

**Comment by jfmengels** [comment]
Open to bikeshedding for this one, for anyone watching the repo :)

I think I prefer pointing to the right tool for each job, as that would directly be more helpful to the user. When you use successive `filter` calls, that would be replaced by `overEvery`. When you use successive `reject` calls, that would be `overSome`. We know the tool he needs to use, let's go and tell them, rather than let them need to look at both of the functions and ponder which one to choose.

**Comment by gunar** [comment]
That's actually what I meant. Didn't make it clear, though. Sounds good! I agree!

**Comment by jfmengels** [comment]
Ok cool, let's go with that then :), will do that tonight

**Comment by jfmengels** [comment]
Landed in 4473a0e
